### PR TITLE
Model confidential and explicit transaction outputs separately

### LINKED ISSUE #66: About contributing back a fairly big change
Hey everyone,

we've vendored this library [here](https://github.com/comit-network/droplet/tree/master/elements-fun) and made some bigger changes across the board, most notably:

- Model `TxOut`s distinctly as explicit or confidential instead of each individual field: https://github.com/comit-network/droplet/blob/8cb45d3029c4edcf499ae8cc21c98e43829a6d59/elements-fun/src/transaction.rs#L79-L90
This model makes expressing nonsensical states like a confidential value with an explicit asset impossible. We've found that quite helpful in trying to wrap our heads around elements. It is quite possible that some edge-cases are not covered but for now all the tests pass.
- Direct integration with our [fork](https://github.com/comit-network/rust-secp256k1-zkp) of `rust-secp256k1` that binds against `libsecp256k1-zkp` to provide all the cryptographic primitives for blinding / unblinding assets etc. This allows us to not have a dependency on `libwally` (which doesn't compile to WASM and has kind of quirky APIs). This is not fully complete (see this [PR](https://github.com/comit-network/droplet/pull/14)) because we need [`SigHashCache`](https://github.com/ElementsProject/rust-elements/pull/64) to land on master here :D 

For now, all of this happens within a single cargo workspace so we don't have to bother with releases. However, we are keen to upstream some of these changes. In particular:

1. Could our [fork](https://github.com/comit-network/rust-secp256k1-zkp) of `rust-secp256k1` potentially replace https://github.com/ElementsProject/rust-secp256k1-zkp? I'd assume we would have to go through some bike-shedding of the APIs as we've currently optimized ours for easy of understanding (YMMV) and to prevent misuse as much as possible at the cost of doing extra allocations for example.
2. The already mentioned refactoring of how `Transaction`s are modeled might be of interest. Similarly, it would take some effort to untangle the changes made there to make them upstreamable because we for example didn't adhere to a MSRV of 1.29.

For example, something that would be possible if we can get (1) done is:

- release a new version of `secp256k1-zkp` to crates.io (it seems that the name `secp256k1-zkp` is pointing to https://github.com/ElementsProject/rust-secp256k1-zkp)
- depend on it in `rust-elements`
- upstream changes like [`ConfidentialTxOut::unblind()`](https://github.com/comit-network/droplet/blob/54f9f4bda551a35d210ee14a57517cc060d47552/elements-fun/src/transaction.rs#L1296-L1333) which provides a quite straight-forward API to invoke some decently heavy cryptography.

Directly depending on `secp256k1-zkp` also removes a lot of duplicated knowledge like how to parse and serialize generators, pedersen commitments, proofs, etc. Additionally, it allows us to use all this cryptography from [WASM](https://github.com/comit-network/droplet/tree/master/waves) which is why we are going through all of this :)

Thoughts?

### LINKED ISSUE #67: Modelling confidential vs explicit transaction outputs
As mentioned [here](https://github.com/ElementsProject/rust-elements/issues/66#issuecomment-744080927), this is a dedicated issue for discussing our attempt at re-modelling confidential and explicit transactions.

We assumed that a `TxOut` can either confidential or explicit - i.e. both the asset and the value are either confidential or explicit - but this turned out to [not be true](https://github.com/ElementsProject/rust-elements/issues/66#issuecomment-743783437).

To be honest, I don't fully understand what a "pre-asset" confidential transaction is? Can I read up on this somewhere to get an overview of what the usecases are and what they look like on-chain?

### PR DESCRIPTION
**Pull request by luckysori**
These are most (all?) of the relevant changes we introduced in our fork of this library (found here: https://github.com/comit-network/waves/tree/master/elements-fun).

The motivation for this PR is to allow users of this library to construct valid confidential transactions, something that used to only be feasible (but quite cumbersome) by also depending on `lib-wally` via https://github.com/RCasatta/wally/.

Closes #66.
Closes #67.
Closes https://github.com/comit-network/waves/issues/130.

CC @thomaseizinger.

### CODE REVIEW

File: src/confidential.rs [hunk 1]
```diff
@@ -324,29 +519,155 @@ impl fmt::Display for Nonce {
...
+            p if p == 0x02 || p == 0x03 => {
+                let bytes = <[u8; 33]>::consensus_decode(&mut d)?;
+                Ok(Nonce::Confidential(
+                    PublicKey::from_slice(&bytes).map_err(secp256k1_zkp::Error::Upstream)?,
+                ))
+            }
+            p => return Err(encode::Error::InvalidConfidentialPrefix(p)),
+        }
+    }
+}
```
Comment (thomaseizinger) [inline src/confidential.rs]: I think we might be able to consolidate these `Decodable` implementations with a macro that is parameterized by:

1. the type you want to construct: `Nonce`
2. the function you need to construct the confidential value: `PublicKey::from_slice`
3. the prefix that identifies the confidential variant: `0x02` and `0x03`

File: src/confidential.rs [hunk 2]
```diff
@@ -324,29 +519,155 @@ impl fmt::Display for Nonce {
...
+            1 => {
+                // ignore prefix when decoding an explicit asset
+                d.consume(1);
+                let explicit = Decodable::consensus_decode(&mut d)?;
+                Ok(Nonce::Explicit(explicit))
+            }
+            p if p == 0x02 || p == 0x03 => {
+                let bytes = <[u8; 33]>::consensus_decode(&mut d)?;
+                Ok(Nonce::Confidential(
+                    PublicKey::from_slice(&bytes).map_err(secp256k1_zkp::Error::Upstream)?,
```
Comment (thomaseizinger) [inline src/confidential.rs:646]: Can you get rid of this `map_err` with a `From` implementation between these two errors?

`?` calls `.from` for you.

File: src/encode.rs [hunk 3]
```diff
@@ -94,7 +94,7 @@ pub trait Encodable {
 /// Data which can be encoded in a consensus-consistent way
 pub trait Decodable: Sized {
     /// Decode an object with a well-defined format
-    fn consensus_decode<D: io::Read>(d: D) -> Result<Self, Error>;
+    fn consensus_decode<D: io::BufRead>(d: D) -> Result<Self, Error>;
```
Comment (thomaseizinger) [inline src/encode.rs:117]: Maybe mention in the commit message why we are doing this: so we can peek at the prefix and still directly deserialize the commitments.

File: src/confidential.rs [hunk 4]
```diff
@@ -17,251 +17,231 @@
...
+            }
+            1 => {
+                // ignore prefix when decoding an explicit value
+                d.consume(1);
+                let explicit = u64::swap_bytes(Decodable::consensus_decode(&mut d)?);
+                Ok(Value::Explicit(explicit))
+            }
+            p if p == 0x08 || p == 0x09 => {
+                let bytes = <[u8; 33]>::consensus_decode(&mut d)?;
+                Ok(Value::Confidential(PedersenCommitment::from_slice(&bytes)?))
```
Comment (thomaseizinger) [inline src/confidential.rs:172]: You could also implement `Decodable` for `PedersenCommitment` and then simply delegate here. With `BufRead`, we can still peek at the prefix and use it to decide on what to delegate to while retaining the prefix in the buffer so the implementation can use it to decode.

File: src/confidential.rs [hunk 5]
```diff
@@ -272,21 +252,188 @@ impl Asset {
...
+    fn consensus_encode<S: io::Write>(&self, mut s: S) -> Result<usize, encode::Error> {
+        match *self {
+            Asset::Null => 0u8.consensus_encode(s),
+            Asset::Explicit(n) => {
+                1u8.consensus_encode(&mut s)?;
+                Ok(1 + n.consensus_encode(&mut s)?)
+            }
+            Asset::Confidential(generator) => {
+                s.write_all(&generator.serialize())?;
+                Ok(33)
```
Comment (thomaseizinger) [inline src/confidential.rs:357]: Same here, you could implement `Encodable` for `Generator`.

File: src/confidential.rs (hunk 1 above)
Comment (thomaseizinger) [reply src/confidential.rs]: Such a macro might be even easier to implement if all the inner types implement `Encodable` / `Decodable`.

File: src/confidential.rs [hunk 6]
```diff
@@ -647,6 +720,68 @@ impl<'de> Deserialize<'de> for Nonce {
     }
 }
 
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+/// Blinding factor used for asset commitments.
```
Comment (thomaseizinger) [inline src/confidential.rs:724]: These comments are usually on top of the derives.

File: src/transaction.rs [hunk 7]
```diff
@@ -373,10 +375,165 @@ impl Decodable for TxOut {
     }
 }
 
+pub enum TxOutError {
```
Comment (thomaseizinger) [inline src/transaction.rs:383]: Does this implement `std::error::Error`?

File: src/transaction.rs [hunk 8]
```diff
@@ -373,10 +375,165 @@ impl Decodable for TxOut {
...
+            secp,
+            1,
+            value_commitment.commitment().expect("confidential value"),
+            value,
+            out_vbf.0,
+            &message.to_bytes(),
+            address.script_pubkey().as_bytes(),
+            shared_secret,
+            0,
+            52,
```
Comment (thomaseizinger) [inline src/transaction.rs:433]: Should we document the use of these constants better? IIRC, we got them out of the documentation of libwally.

cc @apoelstra

File: src/transaction.rs [hunk 9]
```diff
@@ -671,6 +699,93 @@ impl TxOut {
...
+    /// Scriptpubkey
+    pub script_pubkey: Script,
+    /// There should be no such thing as a nonce for an explicit
+    /// output, but Elements will deserialize such a thing and even
+    /// produce it.
+    pub nonce: Nonce,
+}
+serde_struct_impl!(ExplicitTxOut, asset, value, script_pubkey, nonce);
+
+pub struct ConfidentialTxOut {
```
Comment (thomaseizinger) [inline src/transaction.rs:717]: No derives on this one?

**Review by thomaseizinger — COMMENTED** [review]
Really good work, thank you @luckysori!

Left some comments, none of them are really pressing though. Take them as you please :)

File: src/confidential.rs (hunk 6 above)
Comment (luckysori) [reply src/confidential.rs:724]: Ooops.

File: src/confidential.rs (hunk 2 above)
Comment (luckysori) [reply src/confidential.rs:646]: Good call!

File: src/transaction.rs (hunk 8 above)
Comment (luckysori) [reply src/transaction.rs:425]: These come from here, yes: https://github.com/ElementsProject/libwally-core/blob/3aa27c1ba1136be5010808c70ff7e8c0f58f66a9/include/wally_elements.h#L142-143

Comment (luckysori) [reply src/transaction.rs:425]: They're recommended by the authors of libwally, but it would be better to say why they've been chosen if we are to document them.

Comment (apoelstra) [reply src/transaction.rs:425]: The first number, 0, is used to shift the exponent, because CT supports floating-point numbers with an explicit exponent but blind mantissa. Setting this to 0 basically disables the feature, which I think represents a bad privacy/efficiency tradeoff and shouldn't be used.

Setting min-bits to 52 gives you room to represent satoshi values up to 2^52 (45MM BTC, in excess of the total 21MM supply) and was chosen to match the range of a double-width floating point value.

Comment (luckysori) [reply src/transaction.rs:425]: > The first number, 0, is used to shift the exponent, because CT supports floating-point numbers with an explicit exponent but blind mantissa. Setting this to 0 basically disables the feature, which I think represents a bad privacy/efficiency tradeoff and shouldn't be used.

What would be a better default value? And should this API expose an optional argument to control it?

Comment (apoelstra) [reply src/transaction.rs:425]: Sorry, I worded that poorly. We **should** leave it at 0, and we shouldn't expose any way to change it.

**Review by sanket1729 — COMMENTED** [review]
Thanks for the PR. Sorry of the delay in reviewing it :( . Can you split the PR into two Prs? The first PR does everything without introducing the new `ExplcitTxOut` and `ConfidentialTxOut` types. The second PR introduces the modeling of different TxOuts. 

I think the first PR would be ready to be and would not require too much discussion while the second one might require some discussion about the merits.

File: src/confidential.rs [hunk 10]
```diff
@@ -17,251 +17,231 @@
...
-    /// Value is committed
-    Confidential(u8, [u8; 32]),
-}
-impl_confidential_commitment!(Value, u64, 0x08, 0x09, u64::swap_bytes);
+#[cfg(feature = "serde")]
+impl Serialize for Value {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::SerializeSeq;
+
+        let seq_len = if *self == Value::Null { 1 } else { 2 };
```
[the remark this answers is not in the document]
Comment (luckysori) [reply src/confidential.rs:201]: Thanks for this comment! This has highlighted that we had introduced some bugs in the serde code which were not showing up because of some missing tests. I'm adding these tests to the split up, less controversial PR which I'll open soon.

**Comment by luckysori** [comment]
Superseded by #77 and #78.
